core/lib/via_btc_client/src/indexer/parser.rs

```diff
@@ -0,0 +1,321 @@
+use bitcoin::{
+    address::NetworkUnchecked,
+    hashes::Hash,
+    script::{Instruction, PushBytesBuf},
+    secp256k1::XOnlyPublicKey,
+    taproot::Signature as TaprootSignature,
+    Address, Amount, Network, ScriptBuf, Transaction, Txid,
+};
+use zksync_basic_types::H256;
+use zksync_types::{Address as EVMAddress, L1BatchNumber};
+
+use crate::types::{
+    CommonFields, L1BatchDAReference, L1BatchDAReferenceInput, L1ToL2Message, L1ToL2MessageInput,
+    Message, ProofDAReference, ProofDAReferenceInput, ProposeSequencer, ProposeSequencerInput,
+    SystemBootstrapping, SystemBootstrappingInput, ValidatorAttestation, ValidatorAttestationInput,
+    Vote,
+};
+
+// TODO: make this a trait (ViaProtocolParser_V1, ViaProtocolParser_V2, etc)
+pub struct MessageParser {
+    network: Network,
+}
+
+impl MessageParser {
+    pub fn new(network: Network) -> Self {
+        Self { network }
+    }
+
+    pub fn parse_transaction(&self, tx: &Transaction) -> Vec<Message> {
+        tx.input
+            .iter()
+            .filter_map(|input| self.parse_input(input, tx))
+            .collect()
+    }
+
+    fn parse_input(&self, input: &bitcoin::TxIn, tx: &Transaction) -> Option<Message> {
+        let witness = &input.witness;
+        if witness.len() < 3 {
```

Comment:
I guess the reason for this early return is the fact that p2tr input that we are parsing needs to have 3 witness elements, but lets try to be more descriptive here with the comment or extracting the value into the const with the proper name.